### PR TITLE
Sei integration

### LINKED ISSUE #603: Sei integration - investigation
Investigate what's needed to support Sei on WH Connect

- gather technical details
- check wallets
- create tasks under #599

### PR DESCRIPTION
**Pull request by M-Picco**
Sei uses a [token translator contract](https://github.com/wormhole-foundation/example-sei-token-translator/blob/main/src/contract.rs) to convert CW20 tokens to a "native" denomination through the [tokenfactory module](https://github.com/sei-protocol/sei-chain/tree/master/x/tokenfactory). There is an off-chain relayer that listens to transfers inbound to Sei and calls this contract accordingly.

Assets native to the Sei change should go through the normal token bridge process.

### CODE REVIEW

File: wormhole-connect/src/config/mainnet.ts
Comment (M-Picco) [inline wormhole-connect/src/config/mainnet.ts]: I want to replace these decimals with a map indexed by the `Context` enum and have `getTokenDecimals` be used exclusively to retrieve this information, but I feel we should leave it for another PR once either this one or the Aptos PR are merged

File: wormhole-connect/src/config/testnet.ts
Comment (M-Picco) [inline wormhole-connect/src/config/testnet.ts]: Is this value `suiDecimals: 8` right for a USDC asset?

Comment (kev1n-peters) [reply wormhole-connect/src/config/testnet.ts]: Nope, this looks like an issue - it should be `6`

File: wormhole-connect/src/config/mainnet.ts
Comment (kev1n-peters) [reply wormhole-connect/src/config/mainnet.ts]: I agree, we should probably wait for either one of those PRs to be merged and then clean this up

File: sdk/src/config/MAINNET.ts
Comment (anondev2323) [inline sdk/src/config/MAINNET.ts]: You don't need `...CONTRACTS.MAINNET.sei` if you're just overriding the values below

File: sdk/src/config/TESTNET.ts
Comment (anondev2323) [inline sdk/src/config/TESTNET.ts]: If you're assigning `core`, `token_bridge` and `nft_bridge` below, you don't need `...CONTRACTS.TESTNET.sei`

File: sdk/src/config/TESTNET.ts
Comment (anondev2323) [inline sdk/src/config/TESTNET.ts]: Check on this?

File: sdk/src/contexts/abstracts/relayer.ts
Comment (anondev2323) [inline sdk/src/contexts/abstracts/relayer.ts:46]: I think we already have a function for this. Should consider using that

File: sdk/src/contexts/sei/contracts.ts
Comment (anondev2323) [inline sdk/src/contexts/sei/contracts.ts:35]: Maybe we should think more about this contracts class, seems not very useful for a few of the newer chains

File: sdk/src/contexts/sei/index.ts
Comment (anondev2323) [inline sdk/src/contexts/sei/index.ts:348]: Could this not be implemented in the contracts class? Maybe it just returns the address, I think that's how Sui does it

File: sdk/src/contexts/sei/index.ts
Comment (anondev2323) [inline sdk/src/contexts/sei/index.ts:355]: Oh this is different because it's not using the Token Bridge Relayer

File: sdk/src/types.ts
Comment (anondev2323) [inline sdk/src/types.ts:60]: Should it be suggested that an integrator provide a rest endpoint for Sei? (Like we do for RPCs?)

File: wormhole-connect/src/config/testnet.ts
Comment (anondev2323) [reply wormhole-connect/src/config/testnet.ts]: Still shows 8 for `suiDecimals` and `aptosDecimals`

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (anondev2323) [inline wormhole-connect/src/sdk/gasEstimates.ts:87]: Didn't we add a `nativeTokenDecimals` on the network config? I think that's what's happening here anyway...

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (anondev2323) [inline wormhole-connect/src/sdk/gasEstimates.ts]: TODO

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (anondev2323) [inline wormhole-connect/src/sdk/gasEstimates.ts:289]: Same here, should use `nativeTokenDecimals`

File: wormhole-connect/src/sdk/index.ts
Comment (anondev2323) [inline wormhole-connect/src/sdk/index.ts:283]: <3

File: wormhole-connect/src/utils/events.ts
Comment (anondev2323) [inline wormhole-connect/src/utils/events.ts:137]: I've been thinking about just adding `toChainId` onto `txData`

File: wormhole-connect/src/utils/index.ts
Comment (anondev2323) [inline wormhole-connect/src/utils/index.ts:107]: network.nativeTokenDecimals

File: wormhole-connect/src/utils/index.ts
Comment (anondev2323) [inline wormhole-connect/src/utils/index.ts]: typo `hexPrefix`

File: wormhole-connect/src/utils/sei.ts
Comment (anondev2323) [inline wormhole-connect/src/utils/sei.ts]: Should this be in the SDK?

File: wormhole-connect/src/views/Redeem/ExplorerLink.tsx
Comment (anondev2323) [inline wormhole-connect/src/views/Redeem/ExplorerLink.tsx:37]: Seems like a common use case, should it be on the network config?

File: sdk/src/contexts/abstracts/tokenBridge.ts
Comment (anondev2323) [inline sdk/src/contexts/abstracts/tokenBridge.ts:43]: Not sure this type works with Solana

**Review by anondev2323 — CHANGES_REQUESTED** [review]

File: wormhole-connect/src/views/WalletModal.tsx
Comment (kev1n-peters) [inline wormhole-connect/src/views/WalletModal.tsx:272]: `getWalletChainText` is unused now so we should remove it to get rid of the unused warning

File: sdk/src/contexts/sei/index.ts
Comment (M-Picco) [reply sdk/src/contexts/sei/index.ts:348]: I think Sui only implements the `getTokenBridgeRelayer` on the contracts class, which returns a `SuiRelayer` object. The `getBridge`/`mustGetBridge` throw a "not implemented" error.

The sui context retrieves the address the same way as here, but there's no method to wrap it ([example](https://github.com/wormhole-foundation/wormhole-connect/blob/development/sdk/src/contexts/sui/index.ts#L174))

File: sdk/src/contexts/sei/contracts.ts
Comment (M-Picco) [reply sdk/src/contexts/sei/contracts.ts:35]: I tried to avoid returning a single address string, since the intention is to return a class built through a framework that abstracts interactions with the contract.

We could try to define an abstraction of our own and move the related logic from the contexts (e.g. the messages built for Sei transactions) to these abstraction implementations.

File: sdk/src/contexts/abstracts/relayer.ts
Comment (M-Picco) [reply sdk/src/contexts/abstracts/relayer.ts:46]: The relayer payload used (for now) for Sei uses a different structure (it's a json instead of a byte array). I wanted to keep the same `ParsedRelayerPayload` interface for all chains, so I opted to make this the default implementation for the parsing method, which is the one shared by EVMs and Sui, and override with the specifics for Sei.

File: sdk/src/config/TESTNET.ts
Comment (M-Picco) [reply sdk/src/config/TESTNET.ts]: I think tendermint finality is instant, and I haven't seen a consistency level used in the cosmos contracts. I'll check again.

File: wormhole-connect/src/config/mainnet.ts
Comment (kev1n-peters) [inline wormhole-connect/src/config/mainnet.ts]: Shouldn't the decimals be `6` on each chain since it's `6` on Sei or am I missing something?

File: sdk/src/types.ts
Comment (M-Picco) [reply sdk/src/types.ts:60]: You are right. I think that PR was merged after I rebased. I'll do it again and add the warning for rest resources as well.

File: wormhole-connect/src/sdk/index.ts
Comment (kev1n-peters) [inline wormhole-connect/src/sdk/index.ts:325]: I like this

File: wormhole-connect/src/sdk/index.ts
Comment (kev1n-peters) [reply wormhole-connect/src/sdk/index.ts:283]: Good refactor

File: wormhole-connect/src/views/Bridge/GasOptions.tsx
Comment (kev1n-peters) [inline wormhole-connect/src/views/Bridge/GasOptions.tsx]: throught - typo

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (M-Picco) [reply wormhole-connect/src/sdk/gasEstimates.ts:87]: I haven't found any reference to `nativeTokenDecimals` on the repo. Do you mean the `decimals` field?

File: sdk/src/config/TESTNET.ts
Comment (M-Picco) [reply sdk/src/config/TESTNET.ts]: I checked and the wormhole cosmos contract doesn't specify a consistency level (see [here](https://github.com/wormhole-foundation/wormhole/blob/main/cosmwasm/contracts/wormhole/src/contract.rs#L324)), so I think we can assume the finality is instant for cosmos chains (Sei/XPLA/Injective/Terra)

File: wormhole-connect/src/config/mainnet.ts
Comment (M-Picco) [reply wormhole-connect/src/config/mainnet.ts]: You're right, those slipped past

File: sdk/src/contexts/abstracts/tokenBridge.ts
Comment (M-Picco) [reply sdk/src/contexts/abstracts/tokenBridge.ts:43]: I modified the method on the solana context to `arrayify` it before returning it

File: sdk/src/types.ts
Comment (anondev2323) [reply sdk/src/types.ts:60]: There's an rpc validation method that throws an error in the console if the integrator doesn't provide an rpc for a given network.  I think it should throw an error if this isn't provided for sei

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (anondev2323) [reply wormhole-connect/src/sdk/gasEstimates.ts:87]: Hm, it looks like it hasn't been added.  I think it should be added

File: sdk/src/contexts/sei/index.ts
Comment (kev1n-peters) [inline sdk/src/contexts/sei/index.ts]: typos: denominationand, transfering

File: wormhole-connect/src/views/Redeem/ExplorerLink.tsx
Comment (M-Picco) [reply wormhole-connect/src/views/Redeem/ExplorerLink.tsx:37]: A little down below in the file the route is set for addresses/objects. I'd leave it for a smaller PR after this one, just in case.

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (M-Picco) [reply wormhole-connect/src/sdk/gasEstimates.ts:87]: I've left out the decimals rework for a separate PR. I think we can include the `nativeTokenDecimals` to this one as well.

File: sdk/src/types.ts
Comment (M-Picco) [reply sdk/src/types.ts:60]: Refactored and implemented for REST and Sei

File: sdk/src/contexts/sei/index.ts
Comment (M-Picco) [reply sdk/src/contexts/sei/index.ts:355]: Yup. The payload is different. For Sei it's a JSON, while the token bridge relayer uses a byte array with offsets.

File: wormhole-connect/src/utils/sei.ts
Comment (M-Picco) [reply wormhole-connect/src/utils/sei.ts]: It'll muddle up the code a bit, but it makes sense. Transfers out of Sei go through the translator contract, so to keep it consistent the sdk should do the same for inbound transfers.

File: wormhole-connect/src/sdk/gasEstimates.ts
Comment (M-Picco) [reply wormhole-connect/src/sdk/gasEstimates.ts]: Done. Had to do it through the aggregator sdk since the cosmos client requires a wallet as a parameter, even if it's for tx simulation/gas simulation.

File: sdk/src/config/MAINNET.ts
Comment (anondev2323) [inline sdk/src/config/MAINNET.ts:135]: Just comment out this `sei` block as well as any sei tokens.  That should remove it completely from the ui, we can uncomment when we are ready for mainnet

Comment (M-Picco) [reply sdk/src/config/MAINNET.ts:135]: Commented the sei related entries on the `mainnet.ts` file

**Review by anondev2323 — APPROVED** [review]

**Pull request merged**
